### ISSUE TITLE
Unable to Recreate TDM Performance for Half Cheetah

### ISSUE THREAD

**Issue by Random-Word** [vitchyr/rlkit#6]
Hi, I'm trying to evaluate TDM as an algorithm to add to our software and the first step was to replicate the results presented in "TEMPORAL DIFFERENCE MODELS: MODEL-FREE DEEP RL FOR MODEL-BASED CONTROL" (Very impressive paper, by the way). Of the examples in this repository, Half Cheetah looks like it trains the fastest in your paper, so I ran that first. When I run it I don't seem to get the same results you present, but I may just be interpreting things incorrectly.

I created a conda env as per the instructions in the readme, with the exception of gtimer where I removed the version requirement gtimer==1.0.1b5 because that version wasn't available.

When I run examples/tdm/cheetah.py, the initial returns are as follows:

<details>
<summary>Expand Initial Return Data</summary>

> Multitask L1 distance to goal Mean             2.2572
> Multitask L1 distance to goal Std              1.43821
> Multitask L1 distance to goal Max              5.63364
> Multitask L1 distance to goal Min              0.0958731
> Multitask Final L1 distance to goal Mean       1.34996
> Multitask Final L1 distance to goal Std        1.69566
> Multitask Final L1 distance to goal Max        5.46802
> Multitask Final L1 distance to goal Min        0.315033
> Multitask L2 distance to goal Mean             2.2572
> Multitask L2 distance to goal Std              1.43821
> Multitask L2 distance to goal Max              5.63364
> Multitask L2 distance to goal Min              0.0958731
> Multitask Final L2 distance to goal Mean       1.34996
> Multitask Final L2 distance to goal Std        1.69566
> Multitask Final L2 distance to goal Max        5.46802
> Multitask Final L2 distance to goal Min        0.315033
> Multitask Env Rewards Mean                    -2.24934
> Multitask Env Rewards Std                      1.44294
> Multitask Env Rewards Max                     -0.0958731
> Multitask Env Rewards Min                     -5.60402
> Multitask L1 distance to goal Mean             2.2572
> Multitask L1 distance to goal Std              1.43821
> Multitask L1 distance to goal Max              5.63364
> Multitask L1 distance to goal Min              0.0958731
> Multitask Final L1 distance to goal Mean       1.34996
> Multitask Final L1 distance to goal Std        1.69566
> Multitask Final L1 distance to goal Max        5.46802
> Multitask Final L1 distance to goal Min        0.315033
> Multitask L2 distance to goal Mean             2.2572
> Multitask L2 distance to goal Std              1.43821
> Multitask L2 distance to goal Max              5.63364
> Multitask L2 distance to goal Min              0.0958731
> Multitask Final L2 distance to goal Mean       1.34996
> Multitask Final L2 distance to goal Std        1.69566
> Multitask Final L2 distance to goal Max        5.46802
> Multitask Final L2 distance to goal Min        0.315033
> Multitask Env Rewards Mean                    -2.24934
> Multitask Env Rewards Std                      1.44294
> Multitask Env Rewards Max                     -0.0958731
> Multitask Env Rewards Min                     -5.60402
> xvels Mean                                     0.192786
> xvels Std                                      0.55423
> xvels Max                                      1.7305
> xvels Min                                     -2.56708
> Final xvels Mean                              -0.873447
> Final xvels Std                                0.629683
> Final xvels Max                               -0.00983972
> Final xvels Min                               -1.92269
> desired xvels Mean                            -0.242309
> desired xvels Std                              2.53803
> desired xvels Max                              5.45818
> desired xvels Min                             -2.77633
> Final desired xvels Mean                      -0.242309
> Final desired xvels Std                        2.53803
> Final desired xvels Max                        5.45818
> Final desired xvels Min                       -2.77633
> xvel errors Mean                               2.24934
> xvel errors Std                                1.44294
> xvel errors Max                                5.60402
> xvel errors Min                                0.0958731
> Final xvel errors Mean                         1.34996
> Final xvel errors Std                          1.69566
> Final xvel errors Max                          5.46802
> Final xvel errors Min                          0.315033
> QF Loss                                        1.94271
> Policy Loss                                    0.141472
> Raw Policy Loss                                0.141472
> Preactivation Policy Loss                      0
> Q Predictions Mean                            -0.141398
> Q Predictions Std                              0.188076
> Q Predictions Max                             -0.000109524
> Q Predictions Min                             -0.671575
> Q Targets Mean                                -2.11527
> Q Targets Std                                  8.76126
> Q Targets Max                                 -0
> Q Targets Min                                -73.5125
> Bellman Errors Mean                           80.6921
> Bellman Errors Std                           504.045
> Bellman Errors Max                          5365.62
> Bellman Errors Min                             8.74315e-13
> Policy Action Mean                            -0.00243206
> Policy Action Std                              0.00429308
> Policy Action Max                              0.00416228
> Policy Action Min                             -0.011675
> Test Rewards Mean                             -2.24934
> Test Rewards Std                               1.44294
> Test Rewards Max                              -0.0958731
> Test Rewards Min                              -5.60402
> Test Returns Mean                           -222.684
> Test Returns Std                             133.611
> Test Returns Max                             -52.0534
> Test Returns Min                            -534.119
> Test Actions Mean                              0.181862
> Test Actions Std                               0.903677
> Test Actions Max                               1
> Test Actions Min                              -1
> Num Paths                                     10
> Exploration Rewards Mean                    -327.502
> Exploration Rewards Std                      150.894
> Exploration Rewards Max                       -0.0122693
> Exploration Rewards Min                     -684.543
> Exploration Returns Mean                  -32422.7
> Exploration Returns Std                    14259.2
> Exploration Returns Max                    -9411.74
> Exploration Returns Min                   -56577.3
> Exploration Actions Mean                       0.0733828
> Exploration Actions Std                        0.691484
> Exploration Actions Max                        1
> Exploration Actions Min                       -1
> AverageReturn                               -222.684
> Number of train steps total                20075
> Number of env steps total                   1000
> Number of rollouts total                      10
> Train Time (s)                               185.599
> (Previous) Eval Time (s)                       0
> Sample Time (s)                                0.594464
> Epoch Time (s)                               186.193
> Total Train Time (s)                         187.161
> Epoch                                          0

</details>

and the final returns after 27 000 samples were:

<details>
<summary>Expand Final Return Data</summary>

> Multitask L1 distance to goal Mean             2.84896
> Multitask L1 distance to goal Std              1.86101
> Multitask L1 distance to goal Max              8.48643
> Multitask L1 distance to goal Min              0.00163637
> Multitask Final L1 distance to goal Mean       1.12561
> Multitask Final L1 distance to goal Std        1.09752
> Multitask Final L1 distance to goal Max        3.6141
> Multitask Final L1 distance to goal Min        0.010751
> Multitask L2 distance to goal Mean             2.84896
> Multitask L2 distance to goal Std              1.86101
> Multitask L2 distance to goal Max              8.48643
> Multitask L2 distance to goal Min              0.00163637
> Multitask Final L2 distance to goal Mean       1.12561
> Multitask Final L2 distance to goal Std        1.09752
> Multitask Final L2 distance to goal Max        3.6141
> Multitask Final L2 distance to goal Min        0.010751
> Multitask Env Rewards Mean                    -2.83008
> Multitask Env Rewards Std                      1.86414
> Multitask Env Rewards Max                     -0.00163637
> Multitask Env Rewards Min                     -8.48643
> Multitask L1 distance to goal Mean             2.84896
> Multitask L1 distance to goal Std              1.86101
> Multitask L1 distance to goal Max              8.48643
> Multitask L1 distance to goal Min              0.00163637
> Multitask Final L1 distance to goal Mean       1.12561
> Multitask Final L1 distance to goal Std        1.09752
> Multitask Final L1 distance to goal Max        3.6141
> Multitask Final L1 distance to goal Min        0.010751
> Multitask L2 distance to goal Mean             2.84896
> Multitask L2 distance to goal Std              1.86101
> Multitask L2 distance to goal Max              8.48643
> Multitask L2 distance to goal Min              0.00163637
> Multitask Final L2 distance to goal Mean       1.12561
> Multitask Final L2 distance to goal Std        1.09752
> Multitask Final L2 distance to goal Max        3.6141
> Multitask Final L2 distance to goal Min        0.010751
> Multitask Env Rewards Mean                    -2.83008
> Multitask Env Rewards Std                      1.86414
> Multitask Env Rewards Max                     -0.00163637
> Multitask Env Rewards Min                     -8.48643
> xvels Mean                                     0.0265717
> xvels Std                                      1.19521
> xvels Max                                      3.78229
> xvels Min                                     -5.17487
> Final xvels Mean                              -0.953468
> Final xvels Std                                2.0584
> Final xvels Max                                3.35745
> Final xvels Min                               -4.45172
> desired xvels Mean                            -1.27358
> desired xvels Std                              3.19157
> desired xvels Max                              4.8695
> desired xvels Min                             -5.75411
> Final desired xvels Mean                      -1.27358
> Final desired xvels Std                        3.19157
> Final desired xvels Max                        4.8695
> Final desired xvels Min                       -5.75411
> xvel errors Mean                               2.83008
> xvel errors Std                                1.86414
> xvel errors Max                                8.48643
> xvel errors Min                                0.00163637
> Final xvel errors Mean                         1.12561
> Final xvel errors Std                          1.09752
> Final xvel errors Max                          3.6141
> Final xvel errors Min                          0.010751
> QF Loss                                        5.10427
> Policy Loss                                   21.2602
> Raw Policy Loss                               21.2602
> Preactivation Policy Loss                      0
> Q Predictions Mean                           -35.1532
> Q Predictions Std                             79.3634
> Q Predictions Max                             -0.176604
> Q Predictions Min                           -579.39
> Q Targets Mean                               -34.9638
> Q Targets Std                                 80.9855
> Q Targets Max                                 -0.0394544
> Q Targets Min                               -585.102
> Bellman Errors Mean                          134.548
> Bellman Errors Std                           629.529
> Bellman Errors Max                          6193.98
> Bellman Errors Min                             1.83182e-05
> Policy Action Mean                             0.043988
> Policy Action Std                              0.897599
> Policy Action Max                              1
> Policy Action Min                             -1
> Test Rewards Mean                             -2.83008
> Test Rewards Std                               1.86414
> Test Rewards Max                              -0.00163637
> Test Rewards Min                              -8.48643
> Test Returns Mean                           -280.178
> Test Returns Std                             150.009
> Test Returns Max                             -63.9594
> Test Returns Min                            -552.089
> Test Actions Mean                             -0.262005
> Test Actions Std                               0.870809
> Test Actions Max                               1
> Test Actions Min                              -1
> Num Paths                                     10
> Exploration Rewards Mean                    -189.165
> Exploration Rewards Std                      142.202
> Exploration Rewards Max                       -0.217348
> Exploration Rewards Min                     -668.823
> Exploration Returns Mean                  -18727.3
> Exploration Returns Std                    11106.4
> Exploration Returns Max                    -5627.08
> Exploration Returns Min                   -44754
> Exploration Actions Mean                      -0.0129454
> Exploration Actions Std                        0.841991
> Exploration Actions Max                        1
> Exploration Actions Min                       -1
> AverageReturn                               -280.178
> Number of train steps total               670075
> Number of env steps total                  27000
> Number of rollouts total                     272
> Train Time (s)                               339.624
> (Previous) Eval Time (s)                       0.961533
> Sample Time (s)                                0.663905
> Epoch Time (s)                               341.249
> Total Train Time (s)                        8136.35
> Epoch                                         26

</details>

Comparing the final xvel error mean (which I think is the metric you graph in the paper) it seems that after 27 000 env steps, rather than dropping from ~2.5 to < 1, the value has remained at or above 2.5. Is this just a matter of running it for longer, or does this suggest there's something fundamentally wrong with my environment or perhaps that the code has been changed since generating the graphs in the paper?

I've seen similar results in three runs, so it doesn't seem to be a bad initialization.

Thanks for your help.

**Comment by vitchyr** [comment]
Yeah, it's bit hard to know which of those numbers to look at (sorry!), but the correct one is, "Final xvel errors Mean." Looks like it's still rather high. It also seems to start rather low, which is a bit suspicious. I'll double check that the hyperparameters are correct and run the script myself.

**Comment by Random-Word** [comment]
Thanks Vitchyr. It looks like the Ant with goal and velocity targets example is working well.

**Comment by vitchyr** [comment]
No problem! So I checked the hyperparameters, and the only difference with what I had originally ran was that the example had `reward_scale=100` instead of `reward_scale=10`. I updated the example script (e736bfcb9a90ef686d179117faf04e588913e66d) to have the correct reward scale and also to use more rollouts for eval.

After running the script myself, I got a plot of "Final xvel errors Mean vs Epoch (1k steps)" that looks like:
![image](https://user-images.githubusercontent.com/1297125/40456987-00d32588-5ea9-11e8-88de-89772bfed4ba.png)

I'll mark this as closed for now, but feel free to re-open if you have issues.
